### PR TITLE
!!! 4191 - Adjust and test node type change behavior

### LINKED ISSUE #4191: BUG: Changing a NodeType type doesn't get marked as "changed"
If you change the type of a NodeType the workspace doesn't get marked as "changed". So you are not able to publish this particular change without changing something else.

As in the new CR nodes are consistently typed across variants there are some open questions to clarify:
- How shall we mark this change as "changed" without having a dimension space point for this change
- Does it make sence to mark all dimension space points as "changed"?
- How does publishing than work? Do I need to publish each dimension space point or just one to get this change published?
- How do we deal with permissions if a user has e.g. permissions for language A but not for language B?

See also: https://neos-project.slack.com/archives/C3MCBK6S2/p1681472294867779

### PR DESCRIPTION
**Pull request by nezaniel**
Resolves #4191 

This introduces a more comprehensive test suite for ChangeNodeType and adds the following features
* default values are now added if the property is yet missing
* !!! obsolete properties are now removed
* missing tethered children are now added
* disallowed children are now properly removed (in DELETE strategy)
* already present tethered children will be adjusted in type if necessary
* all the above is now applied recursively for nested tethered declarations

This also fixes a bug in NodeRemoval where only one child node was removed per removed parent

Additionally, as reported in #4191, the change projection is now able to deal with aggregate scoped changes (NodeAggregateTypeWasChanged, NodeAggregateNameWasChanged)

**Upgrade instructions**

Requires cr:setup

**Review instructions**

See change list

**Checklist**

- [X] Code follows the PSR-12 coding style
- [X] Tests have been created, run and adjusted as needed
- [X] The PR is created against the 9.0 branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/PendingChangesProjection/Change.php [hunk 1]
```diff
@@ -55,8 +56,8 @@ public function addToDatabase(Connection $databaseConnection, string $tableName)
             $databaseConnection->insert($tableName, [
                 'contentStreamId' => $this->contentStreamId->value,
                 'nodeAggregateId' => $this->nodeAggregateId->value,
-                'originDimensionSpacePoint' => $this->originDimensionSpacePoint->toJson(),
-                'originDimensionSpacePointHash' => $this->originDimensionSpacePoint->hash,
+                'originDimensionSpacePoint' => $this->originDimensionSpacePoint?->toJson(),
+                'originDimensionSpacePointHash' => $this->originDimensionSpacePoint?->hash ?: 'AGGREGATE',
```
Comment (dlubitz) [inline Neos.Neos/Classes/PendingChangesProjection/Change.php]: Why do we need this string? Can't we change the column to nullable as we have for the DSP? And if not, can we put it into a constant?

File: Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php [hunk 2]
```diff
@@ -148,4 +160,206 @@ protected function createEventsForMissingTetheredNode(
...
+                        $currentNodePath->appendPathSegment($childTetheredNodeTypeDefinition->name),
+                    )
+                )
+            );
+        }
+
+        return Events::fromArray($events);
+    }
+
+    protected function createEventsForWronglyTypedNodeAggregate(
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php:260]: I can't see where this is ever called? It calls itself recursively, but outside of that?

File: Neos.Neos/Classes/PendingChangesProjection/Change.php (hunk 1 above)
Comment (kitsunet) [reply Neos.Neos/Classes/PendingChangesProjection/Change.php]: What I do not undesrtand, here we use this fallback string, but down below we build a query against NULL values?

Comment (dlubitz) [reply Neos.Neos/Classes/PendingChangesProjection/Change.php]: Yes, that's correct (it's correct -> this was not correct). We already fixed that.

File: Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php (hunk 2 above)
Comment (dlubitz) [reply Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php:260]: It's called here:
https://github.com/neos/neos-development-collection/blob/c8f13d35ac1ffce6eb327c67418459d5dae79ac6/Neos.ContentRepository.Core/Classes/Feature/NodeTypeChange/NodeTypeChange.php#L268-L268

File: Neos.Neos/Classes/PendingChangesProjection/Change.php (hunk 1 above)
Comment (dlubitz) [reply Neos.Neos/Classes/PendingChangesProjection/Change.php]: As @nezaniel stated out: We need to set something here, as this column is part of the primary index and can't be empty.

File: Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php (hunk 2 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/Common/TetheredNodeInternals.php:260]: Ah thanks!

Seach why would you fail me like this 😆

File: Neos.Neos/Classes/PendingChangesProjection/Change.php [hunk 3]
```diff
@@ -83,8 +86,8 @@ public function updateToDatabase(Connection $databaseConnection, string $tableNa
                 [
                     'contentStreamId' => $this->contentStreamId->value,
                     'nodeAggregateId' => $this->nodeAggregateId->value,
-                    'originDimensionSpacePoint' => $this->originDimensionSpacePoint->toJson(),
-                    'originDimensionSpacePointHash' => $this->originDimensionSpacePoint->hash,
+                    'originDimensionSpacePoint' => $this->originDimensionSpacePoint?->toJson(),
+                    'originDimensionSpacePointHash' => $this->originDimensionSpacePoint?->hash ?: 'AGGREGATE',
```
Comment (kitsunet) [inline Neos.Neos/Classes/PendingChangesProjection/Change.php]:
```suggestion
                    'originDimensionSpacePointHash' => $this->originDimensionSpacePoint?->hash ?: self::AGGREGATE_DIMENSIONSPACEPOINT_HASH_PLACEHOLDER,
```

**Review by kitsunet — APPROVED** [review]

**Review by dlubitz — APPROVED** [review]

File: Neos.Neos/Classes/PendingChangesProjection/Change.php [hunk 4]
```diff
@@ -30,13 +30,16 @@
...
+
     /**
      * @param NodeAggregateId|null $removalAttachmentPoint {@see RemoveNodeAggregate::$removalAttachmentPoint} for docs
      */
     public function __construct(
         public ContentStreamId $contentStreamId,
         public NodeAggregateId $nodeAggregateId,
-        public OriginDimensionSpacePoint $originDimensionSpacePoint,
+        // null for aggregate scoped changes (e.g. NodeAggregateNameWasChanged, NodeAggregateTypeWasChanged)
+        public ?OriginDimensionSpacePoint $originDimensionSpacePoint,
```
Comment (mhsdesign) [inline Neos.Neos/Classes/PendingChangesProjection/Change.php:42]: todo the neos ui must learn to handle this case: https://github.com/neos/neos-ui/blob/13520defa0b00eee9cef6073fde11ccbfa3346d4/Classes/ContentRepository/Service/WorkspaceService.php#L83

i assume we need to mark all occupied nodes of the noteaggregate as changes then? And if `removalAttachmentPoint` is set is probably gurantted that `originDimensionSpacePoint` is set

Comment (mhsdesign) [reply Neos.Neos/Classes/PendingChangesProjection/Change.php:42]: okay it _did_ see https://github.com/neos/neos-ui/pull/3878 ~but one place was overseen :)~

**Pull request merged**